src/datomish/clauses.cljc

```diff
@@ -0,0 +1,256 @@
+
+;; `from` is a list of [source alias] pairs, suitable for passing to honeysql.
+;; `bindings` is a map from var to qualified columns.
+;; `wheres` is a list of fragments that can be joined by `:and`.
+(defrecord ConjoiningClauses [source from external-bindings bindings wheres])
+
+(defn bind-column-to-var [cc variable col]
+  (let [var (:symbol variable)]
+    (util/conj-in cc [:bindings var] col)))
+
+(defn constrain-column-to-constant [cc col position value]
+  (util/conj-in cc [:wheres]
+                [:= col (if (= :a position)
+                          (attribute-in-source (:source cc) value)
+                          (constant-in-source (:source cc) value))]))
+
+(defn- bindings->where
+  "Take a bindings map like
+    {?foo [:datoms12.e :datoms13.v :datoms14.e]}
+  and produce a list of constraints expression like
+    [[:= :datoms12.e :datoms13.v] [:= :datoms12.e :datoms14.e]]
+
+  TODO: experiment; it might be the case that producing more
+  pairwise equalities we get better or worse performance."
+  [bindings]
+  (mapcat (fn [[_ vs]]
+            (when (> (count vs) 1)
+              (let [root (first vs)]
+                (map (fn [v] [:= root v]) (rest vs)))))
+          bindings))
+
+;; This is so we can link clauses to the outside world.
+(defn impose-external-bindings [cc]
+  (if (empty? (:external-bindings cc))
+    cc
+    (let [ours (:bindings cc)
+          theirs (:external-bindings cc)
+          vars (clojure.set/intersection (set (keys theirs)) (set (keys ours)))]
+      (util/concat-in
+        cc [:wheres]
+        (map
+          (fn [v]
+            (let [external (first (v theirs))
+                  internal (first (v ours))]
+              (assert external)
+              (assert internal)
+              [:= external internal]))
+          vars)))))
+
+(defn expand-where-from-bindings
+  "Take the bindings in the CC and contribute
+   additional where clauses. Calling this more than
+   once will result in duplicate clauses."
+  [cc]
+  (impose-external-bindings
+    (assoc cc :wheres (concat (bindings->where (:bindings cc))
+                              (:wheres cc)))))
+
+;; Pattern building is recursive, so we need forward declarations.
+(declare Not->NotJoinClause not-join->where-fragment)
+
+;; Accumulates a pattern into the CC. Returns a new CC.
+(defn apply-pattern-clause
+  "Transform a DataScript Pattern instance into the parts needed
+   to build a SQL expression.
+
+  @arg cc A CC instance.
+  @arg pattern The pattern instance.
+  @return an augmented CC"
+  [cc pattern]
+  (when-not (instance? Pattern pattern)
+    (raise-str "Expected to be called with a Pattern instance." pattern))
+  (when-not (instance? DefaultSrc (:source pattern))
+    (raise-str "Non-default sources are not supported in patterns. Pattern: " pattern))
+
+  (let [[table alias] (source->from (:source cc))   ; e.g., [:datoms :datoms123]
+        places (map (fn [place col] [place col])
```

Comment:
It's not identity — it's a two-argument function that returns a two-element array.

```
user=> (map identity [1 2 3] [4 5 6])

ArityException Wrong number of args (2) passed to: core/identity  clojure.lang.AFn.throwArity (AFn.java:429)
```

We could use `clojure.core.logic/->Pair`, but I suspect that's overkill :)